parley/src/layout/line/greedy.rs

```diff
@@ -51,196 +117,248 @@ impl<'a, B: Brush> BreakLines<'a, B> {
+        self.state.items = self.lines.line_items.len();
+        self.state.lines = self.lines.lines.len();
+        self.state.line.x = 0.;
+        self.state.prev_boundary = None; // Added by Nico
+        self.last_line_data()
+    }
+
+    fn last_line_data(&self) -> Option<(f32, f32)> {
+        let line = self.lines.lines.last().unwrap();
+        Some((line.metrics.advance, line.size()))
+    }
+
     /// Computes the next line in the paragraph. Returns the advance and size
     /// (width and height for horizontal layouts) of the line.
     pub fn break_next(&mut self, max_advance: f32, alignment: Alignment) -> Option<(f32, f32)> {
+        // Maintain iterator state
         if self.done {
             return None;
         }
         self.prev_state = Some(self.state.clone());
-        let run_count = self.layout.runs.len();
-        while self.state.i < run_count {
-            let run_data = &self.layout.runs[self.state.i];
-            let run = Run::new(self.layout, run_data, None);
-            let cluster_start = run_data.cluster_range.start;
-            let cluster_end = run_data.cluster_range.end;
-            while self.state.j < cluster_end {
-                let cluster = run.get(self.state.j - cluster_start).unwrap();
-                let is_ligature_continuation = cluster.is_ligature_continuation();
-                let is_space = cluster.info().whitespace().is_space_or_nbsp();
-                let boundary = cluster.info().boundary();
-                match boundary {
-                    Boundary::Mandatory => {
-                        if !self.state.line.skip_mandatory_break {
-                            self.state.prev_boundary = None;
-                            self.state.line.clusters.end = self.state.j;
-                            self.state.line.runs.end = self.state.i + 1;
-                            self.state.line.skip_mandatory_break = true;
-                            if commit_line(
-                                self.layout,
-                                &mut self.lines,
-                                &mut self.state.line,
-                                max_advance,
-                                alignment,
-                                BreakReason::Explicit,
-                                false,
-                            ) {
-                                self.state.runs = self.lines.runs.len();
-                                self.state.lines = self.lines.lines.len();
-                                self.state.line.x = 0.;
-                                let line = self.lines.lines.last().unwrap();
-                                return Some((line.metrics.advance, line.size()));
+
+        // This macro simply calls the `commit_line` with the provided arguments and some parts of self.
+        // It exists solely to cut down on the boilerplate for accessing the self variables while
+        // keeping the borrow checker happy
+        macro_rules! try_commit_line {
```

Comment:
I think the macro and item namespaces are different, so you should be able to use the same name as the function.
If the function should be named `try_commit_line`, then maybe that change should be made

Reply:
Function renamed to `try_commit_line` which I think is a better name as it has an early return that doesn't commit line (this is also something for which I'm not sure of the reason, but removing this broke things (caused panics)).